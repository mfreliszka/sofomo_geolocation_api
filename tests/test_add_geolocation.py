"""Module containing tests for the get_geolocation endpoint."""

import json
from fastapi import status
import pytest

from syrupy.assertion import SnapshotAssertion


@pytest.mark.anyio
async def test_add_new_geolocation(
    app: pytest.fixture,
    snapshot: SnapshotAssertion,
    valid_ip1: pytest.fixture,
    httpx_async_client: pytest.fixture,
):
    """Test adding a new geolocation to the database."""
    response = await httpx_async_client.post(
        app.url_path_for("add_geolocation_to_database"),
        content=json.dumps(dict({"ip_address": valid_ip1})),
        headers={"Content-Type": "application/json"},
    )

    assert response.json() == snapshot, "Actual response did not match snapshot"
    assert response.status_code == status.HTTP_201_CREATED, (
        f"Response status code: {response.status_code}"
    )
    assert response.headers["Content-Type"] == "application/json", (
        f"Response content type: {response.headers['Content-Type']}"
    )


@pytest.mark.anyio
async def test_add_existing_geolocation(
    app: pytest.fixture,
    IPGeolocation1_InDB_Model: pytest.fixture,
    snapshot: SnapshotAssertion,
    valid_ip1: pytest.fixture,
    httpx_async_client: pytest.fixture,
):
    """Test adding an existing geolocation to the database."""
    response = await httpx_async_client.post(
        app.url_path_for("add_geolocation_to_database"),
        content=json.dumps(dict({"ip_address": valid_ip1})),
        headers={"Content-Type": "application/json"},
    )

    assert response.json() == snapshot, "Actual response did not match snapshot"
    assert response.status_code == status.HTTP_200_OK, (
        f"Response status code: {response.status_code}"
    )
    assert response.headers["Content-Type"] == "application/json", (
        f"Response content type: {response.headers['Content-Type']}"
    )


@pytest.mark.anyio
async def test_add_invalid_ip_address(
    app: pytest.fixture,
    snapshot: SnapshotAssertion,
    invalid_ip: pytest.fixture,
    httpx_async_client: pytest.fixture,
):
    """Test adding invalid IP address to the database."""
    response = await httpx_async_client.post(
        app.url_path_for("add_geolocation_to_database"),
        content=json.dumps(dict({"ip_address": invalid_ip})),
        headers={"Content-Type": "application/json"},
    )

    assert response.json() == snapshot, "Actual response did not match snapshot"
    assert response.status_code == status.HTTP_400_BAD_REQUEST, (
        f"Response status code: {response.status_code}"
    )
    assert response.headers["Content-Type"] == "application/json", (
        f"Response content type: {response.headers['Content-Type']}"
    )
