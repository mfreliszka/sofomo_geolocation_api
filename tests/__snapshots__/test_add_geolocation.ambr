# serializer version: 1
# name: test_add_existing_geolocation[asyncio]
  dict({
    'data': dict({
      'city': 'Mountain View',
      'continent_code': 'NA',
      'continent_name': 'North America',
      'country_code': 'US',
      'country_name': 'United States',
      'id': 1,
      'ip': '8.8.8.8',
      'latitude': 37.422,
      'longitude': -122.084,
      'region_code': 'CA',
      'region_name': 'California',
      'type': 'IPv4',
      'updated_at': '2020-01-01T00:00:00',
      'zip': '94043',
    }),
    'message': "Geolocation for IP '8.8.8.8' already exists in database",
  })
# ---
# name: test_add_existing_geolocation[trio]
  dict({
    'data': dict({
      'city': 'Mountain View',
      'continent_code': 'NA',
      'continent_name': 'North America',
      'country_code': 'US',
      'country_name': 'United States',
      'id': 1,
      'ip': '8.8.8.8',
      'latitude': 37.422,
      'longitude': -122.084,
      'region_code': 'CA',
      'region_name': 'California',
      'type': 'IPv4',
      'updated_at': '2020-01-01T00:00:00',
      'zip': '94043',
    }),
    'message': "Geolocation for IP '8.8.8.8' already exists in database",
  })
# ---
# name: test_add_invalid_ip_address[asyncio]
  dict({
    'detail': "Invalid IP address: 'invalid_ip'",
  })
# ---
# name: test_add_invalid_ip_address[trio]
  dict({
    'detail': "Invalid IP address: 'invalid_ip'",
  })
# ---
# name: test_add_new_geolocation[asyncio]
  dict({
    'city': 'Glenmont',
    'continent_code': 'NA',
    'continent_name': 'North America',
    'country_code': 'US',
    'country_name': 'United States',
    'id': 1,
    'ip': '8.8.8.8',
    'latitude': 40.5369987487793,
    'longitude': -82.1285934448242,
    'region_code': 'OH',
    'region_name': 'Ohio',
    'type': 'ipv4',
    'updated_at': '2020-01-01T00:00:00',
    'zip': '44628',
  })
# ---
# name: test_add_new_geolocation[trio]
  dict({
    'city': 'Glenmont',
    'continent_code': 'NA',
    'continent_name': 'North America',
    'country_code': 'US',
    'country_name': 'United States',
    'id': 1,
    'ip': '8.8.8.8',
    'latitude': 40.5369987487793,
    'longitude': -82.1285934448242,
    'region_code': 'OH',
    'region_name': 'Ohio',
    'type': 'ipv4',
    'updated_at': '2020-01-01T00:00:00',
    'zip': '44628',
  })
# ---
